src/org/jitsi/impl/neomedia/transform/RetransmissionRequesterDelegate.java

```diff
@@ -0,0 +1,582 @@
+
+    /**
+     * A callback which allows this class to signal it has nack work that is ready
+     * to be run
+     */
+    protected Runnable workReadyCallback = null;
+
+    /**
+     * Initializes a new <tt>RetransmissionRequesterDelegate</tt> for the given
+     * <tt>RtpChannel</tt>.
+     * @param stream the {@link MediaStream} that the instance belongs to.
+     */
+    public RetransmissionRequesterDelegate(MediaStream stream, TimeProvider timeProvider)
+    {
+        super(RTPPacketPredicate.INSTANCE);
+        this.stream = stream;
+        this.timeProvider = timeProvider;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Implements {@link SinglePacketTransformer#reverseTransform(RawPacket)}.
+     */
+    @Override
+    public RawPacket reverseTransform(RawPacket pkt)
+    {
+        Long ssrc;
+        int seq;
+
+        MediaFormat format = stream.getFormat(pkt.getPayloadType());
+        if (format == null)
+        {
+            ssrc = null;
+            seq = -1;
+
+            logger.warn("format_not_found" +
+                ",stream_hash=" + stream.hashCode());
+        }
+        else if (Constants.RTX.equalsIgnoreCase(format.getEncoding()))
+        {
+            MediaStreamTrackReceiver receiver
+                = stream.getMediaStreamTrackReceiver();
+
+            RTPEncodingDesc encoding = receiver.findRTPEncodingDesc(pkt);
+
+            if (encoding != null)
+            {
+                ssrc = encoding.getPrimarySSRC();
+                seq = pkt.getOriginalSequenceNumber();
+            }
+            else
+            {
+                ssrc = null;
+                seq = -1;
+
+                logger.warn("encoding_not_found" +
+                    ",stream_hash=" + stream.hashCode());
+            }
+        }
+        else
+        {
+            ssrc = pkt.getSSRCAsLong();
+            seq = pkt.getSequenceNumber();
+        }
+
+
+        if (ssrc != null)
+        {
+            // TODO(gp) Don't NACK higher temporal layers.
+            Requester requester = getOrCreateRequester(ssrc);
+            requester.received(seq);
+        }
+
+        return pkt;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public long getTimeUntilNextRun()
+    {
+        long now = timeProvider.getTime();
+        List<Requester> dueRequesters = getDueRequesters(now);
+        if (dueRequesters.isEmpty())
+        {
+            return WAKEUP_INTERVAL_MILLIS;
+        }
+        else
+        {
+            if (logger.isTraceEnabled())
+            {
+                Requester nextDueRequester = dueRequesters.get(0);
+                logger.trace(hashCode() + ": Next nack is scheduled for ssrc " +
+                    nextDueRequester.ssrc + " at " +
+                    Math.max(nextDueRequester.nextRequestAt, 0) +
+                    ".  (current time is " + now + ")");
+            }
+            return Math.max(dueRequesters.get(0).nextRequestAt - now, 0);
```

Comment:
I don't see getDueRequesters doing anything about the order of the requesters it receives, and I think here we want to min across all requesters. I would also add a method that just iterates and returns the minimum as an int to avoid creating a list every time.

Reply:
yep you're right, i was sorting this at one point but i must've lost it amongst some other changes.  will fix that.

Reply:
i ended up adding a new method `getNextDueRequester` which is used by `getTimeUntilNextRun` and returns the requester (if any) which is due next. The other usage of `getDueRequesters` is the `run` method which doesn't actually care about the order so i left that method as it was.  there's a little duplication there...could have one method that just returned the sorted list, but this way as now we don't build the list every time we want to find out the next time...if you think it'd be better this way i'm fine with it.